plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/ConnectionWrapper.java

```diff
@@ -0,0 +1,244 @@
+import org.jboss.tools.openshift.internal.core.WatchManager;
+import org.jboss.tools.openshift.internal.ui.OpenShiftUIActivator;
+
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class ConnectionWrapper extends AbstractOpenshiftUIElement<IOpenShiftConnection, OpenshiftUIModel> {
+	public static final String[] RESOURCE_KINDS = { ResourceKind.BUILD, ResourceKind.BUILD_CONFIG,
+			ResourceKind.DEPLOYMENT_CONFIG, ResourceKind.IMAGE_STREAM, ResourceKind.IMAGE_STREAM_TAG, ResourceKind.POD,
+			ResourceKind.ROUTE, ResourceKind.REPLICATION_CONTROLLER, ResourceKind.SERVICE };
+
+	private AtomicReference<LoadingState> state = new AtomicReference<LoadingState>(LoadingState.INIT);
+	private Map<IProject, ProjectWrapper> projects = new HashMap<>();
+	private ResourceCache resourceCache = new ResourceCache();
+
+	public ConnectionWrapper(OpenshiftUIModel parent, IOpenShiftConnection wrapped) {
+		super(parent, wrapped);
+	}
+
+	public IOpenShiftConnection getConnection() {
+		return getWrapped();
+	}
+
+	public Collection<ProjectWrapper> getProjects() {
+		synchronized (projects) {
+			return new ArrayList<ProjectWrapper>(projects.values());
+		}
+	}
+
+	public LoadingState getState() {
+		return state.get();
+	}
+	
+	void initWith(List<IResource> resources) {
+		synchronized (projects) {
+			resources.forEach(project -> {
+				projects.put((IProject) project, new ProjectWrapper(this, (IProject) project));
+			});
+			state.set(LoadingState.LOADED);
+		}
+	}
+
+	public boolean load(IExceptionHandler handler) {
+		if (state.compareAndSet(LoadingState.INIT, LoadingState.LOADING)) {
+			startLoadJob(handler);
+			return true;
+		}
+		return false;
+	}
+
+	void startLoadJob(ProjectWrapper projectWrapper, IExceptionHandler handler) {
+		new Job("load project") {
+
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					WatchManager.getInstance().startWatch(projectWrapper.getResource());
+					Collection<IResource> resources = new HashSet<>();
+					for (String kind : RESOURCE_KINDS) {
+						resources.addAll(projectWrapper.getResource().getResources(kind));
+					}
+					resources.forEach(r -> resourceCache.add(r));
+					projectWrapper.initWithResources(resources);
+					projectWrapper.fireChanged();
+				} catch (OperationCanceledException e) {
+					projectWrapper.setLoadingState(LoadingState.LOAD_STOPPED);
+				} catch (Throwable e) {
+					projectWrapper.setLoadingState(LoadingState.LOAD_STOPPED);
+					handler.handleException(e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+
+	void startLoadJob(IExceptionHandler handler) {
+		new Job("load project") {
+
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					IOpenShiftConnection connection = getConnection();
+					List<IResource> projects = connection.getResources(ResourceKind.PROJECT);
+					initWith(projects);

```

Comment:
Can we have a race condition here? I see we set the state to loaded but I don't see any other locks.  Could we have multiple threads loading at the same time?